dcicutils/misc_utils.py

```diff
@@ -325,3 +325,100 @@ def apply_dict_overrides(dictionary: dict, **overrides) -> dict:
 
 def utc_today_str():
     return datetime.datetime.strftime(datetime.datetime.utcnow(), "%Y-%m-%d")
+
+
+class LockoutManager:
+
+    EARLIEST_TIMESTAMP = datetime.datetime(datetime.MINYEAR, 1, 1)  # maybe useful for testing
+
+    def __init__(self, *, lockout_seconds, safety_seconds=0, action="metered action", enabled=True, log=None):
+        # This makes it easy to turn off the feature
+        self.lockout_enabled = enabled
+        self.lockout_seconds = lockout_seconds
+        self.safety_seconds = safety_seconds
+        self.action = action
+        self._timestamp = self.EARLIEST_TIMESTAMP
+        self.log = log
+
+    @property
+    def timestamp(self):
+        """The timestamp is read-only. Use update_timestamp() to set it."""
+        return self._timestamp
+
+    @property
+    def effective_lockout_seconds(self):
+        return self.lockout_seconds + self.safety_seconds
+
+    def wait_if_needed(self):
+        now = datetime.datetime.now()
+        # Note that this quantity is always positive because now is always bigger than the timestamp.
+        seconds_since_last_purge = (now - self._timestamp).total_seconds()
+        # Note again that because seconds_since_last_attempt is positive, the wait seconds will
+        # never exceed self.effective_lockout_seconds, so
+        #   0 <= wait_seconds <= self.effective_lockout_seconds
+        wait_seconds = max(0.0, self.effective_lockout_seconds - seconds_since_last_purge)
+        if wait_seconds > 0.0:
+            shared_message = ("Last %s attempt was at %s (%s seconds ago)."
+                              % (self.action, self._timestamp, seconds_since_last_purge))
+            if self.lockout_enabled:
+                action_message = "Waiting %s seconds before attempting another." % wait_seconds
+                if self.log:
+                    self.log.warning("%s %s" % (shared_message, action_message))
```

Comment:
I notice you do `if self.log` in some places here, wouldn't it be better to just have a helper method like `log_if_you_can`? Moreover, it feels like these two classes are very similar - can they be refactored to extend a base class? Is it worthwhile to do so?

Reply:
Theoretically one can be a special case of the other, where its `allowed_attempts=1` and the length of the interval is similar to the lockout time. The thing is that the n=1 case has an ability to ask about its timestamp in a way that you _could_ generalize to the multiple case, but it's tricky. On the other hand, I will do the experiment of just using a RateManager on the purge code and see if that makes any difference. The methods like timestamp mostly affect how we unit-test things. Probably if the unit tests were rewritten, it wouldn't matter. After all, the RateManager has its own unit tests.

Reply:
As for the `.log` that's slightly different. The purpose is to be a proxy for the other log. Theoretically, as a library, you don't know how logging is accomplished and you can't just find the logger because that's a lexical matter. So someone does a 
```
LOG = some_log_factory.make_some_log(__file__)
```
kind of call that sets up a log specific to this file and specific to its needs, and while you can make a logger that logs _somewhere_, it's not possible to guess the right place.

That's a separate question from whether I should have done:
```
def warn(*args, **kwargs):
    if self.log:
        self.log.warning(*args, **kwargs)
```
or
```
import logging
...
class LockoutManager:
    DEFAULT_LOGGER = logging
    def __init__(..., log=None):
        self.log = log or DEFAULT_LOGGER
```
The latter seems worth doing, I think. That would give the default effect of your `log_if_you_can` by making sure you always can, and then one could use `self.log.warning(...)` without conditionals.